common/persistence/cassandra/queue_store_v2.go

```diff
@@ -0,0 +1,356 @@
+
+func (q *QueueStoreV2) RangeDeleteMessages(ctx context.Context, request persistence.InternalRangeDeleteMessagesRequest) (*persistence.InternalRangeDeleteMessagesResponse, error) {
+	ackLevel, version, err := q.getCurrentAckLevelAndVersion(ctx, request.QueueType, request.QueueName)
+	if err != nil {
+		return nil, err
+	}
+
+	query := q.session.Query(templateDeleteMessagesQueryV2, request.QueueType, request.QueueName, ackLevel, request.InclusiveMaxMessageMetadata.ID).WithContext(ctx)
+	if err := query.Exec(); err != nil {
+		return nil, serviceerror.NewUnavailable(fmt.Sprintf("RangeDeleteMessagesFromDLQ operation failed. Error %v, Type: %v, Name: %v", err, request.QueueType, request.QueueName))
+	}
+
+	return &persistence.InternalRangeDeleteMessagesResponse{}, q.updateAckLevel(ctx, request, version)
+}
+
+func (q *QueueStoreV2) ListQueues(ctx context.Context, request persistence.InternalListQueuesRequest) (*persistence.InternalListQueuesResponse, error) {
+	query := q.session.Query(templateGetQueuesListQueryV2, request.QueueType).WithContext(ctx)
+	iter := query.PageSize(request.PageSize).PageState(request.NextPageToken).Iter()
+
+	response := persistence.InternalListQueuesResponse{}
+	queueName := make(map[string]interface{})
+	for iter.MapScan(queueName) {
+		response.QueueNames = append(response.QueueNames, queueName["queue_name"].(string))
+		queueName = make(map[string]interface{})
+	}
+
+	if len(iter.PageState()) > 0 {
+		response.NextPageToken = iter.PageState()
+	}
+
+	if err := iter.Close(); err != nil {
+		return nil, serviceerror.NewUnavailable(fmt.Sprintf("ListQueues operation failed. Error: %v, Type: %v", err, request.QueueType))
+	}
+
+	return &response, nil
+}
+
+func (q *QueueStoreV2) Close() {
+	if q.session != nil {
+		q.session.Close()
+	}
+}
+
+func (q *QueueStoreV2) tryEnqueue(
+	ctx context.Context,
+	request persistence.InternalEnqueueMessageRequest,
+	messageID int64,
+	partition int,
+) (int64, error) {
+	query := q.session.Query(templateEnqueueMessageQuery, request.QueueType, request.QueueName, partition, messageID, request.Blob.Data, request.Blob.EncodingType.String()).WithContext(ctx)
+	previous := make(map[string]interface{})
+	applied, err := query.MapScanCAS(previous)
+	if err != nil {
+		return persistence.EmptyQueueMessageID, gocql.ConvertError("tryEnqueue", err)
+	}
+
+	if !applied {
+		return persistence.EmptyQueueMessageID, &persistence.ConditionFailedError{Msg: fmt.Sprintf("message ID %v exists in queue", previous["message_id"])}
+	}
+
+	return messageID, nil
+}
+
+func (q *QueueStoreV2) getLastMessageID(
+	ctx context.Context,
+	request persistence.InternalEnqueueMessageRequest,
+	partition int,
+) (int64, error) {
+	query := q.session.Query(templateGetLastMessageIDQuery, request.QueueType, request.QueueName, partition).WithContext(ctx)
+	result := make(map[string]interface{})
+	err := query.MapScan(result)
+	if err != nil {
+		if gocql.IsNotFoundError(err) {
+			return persistence.EmptyQueueMessageID, nil
```

Comment:
If ack level starts from 0, it means the first valid messageID is 1, so here it needs to return 0 not -1?

Reply:
Good catch, let me fix it.